### DISCUSSION TITLE
Fixing the right-to-left language issue

### DISCUSSION THREAD

**Discussion by Theamoz** [machinewrapped/llm-subtrans#201] [Ideas]
![{3D1E2624-4ABD-452C-BE18-9E3F717D8887}](https://github.com/user-attachments/assets/e47f5016-5473-4368-97c9-0d0fcb38814d)
For example, all symbols instead of the end of the line go to the beginning of the line.
![{6481AE02-FF22-4A19-A417-CABB35C67FCC}](https://github.com/user-attachments/assets/2e6fc41c-c333-4475-835a-c91fb3eda478)
And in this example, where an English word was added to the text, the entire sentence became scrambled and ruined.

I have successfully resolved this issue by implementing this code. It is strongly recommended to incorporate it into your application to ensure proper functionality for right-to-left languages such as Arabic and Persian

```
import pysrt
from collections import Counter
import unicodedata as ud

def dominant_strong_direction(s):
    count = Counter([ud.bidirectional(c) for c in list(s)])
    rtl_count = count['R'] + count['AL'] + count['RLE'] + count["RLI"]
    ltr_count = count['L'] + count['LRE'] + count["LRI"]
    return "rtl" if rtl_count > ltr_count else "ltr"

def convert_srt_to_rtl(input_file, output_file):
    subs = pysrt.open(input_file)
    for sub in subs:
        if dominant_strong_direction(sub.text) == "rtl":
            sub.text = "\u202b" + sub.text + "\u202c"
    subs.save(output_file, encoding='utf-8')

if __name__ == "__main__":
    input_file = "subtitle.srt"
    output_file = "subtitle_rtl.srt"
    convert_srt_to_rtl(input_file, output_file)
```
These are two examples that show the issue has been successfully resolved and also works well for all right-to-left languages
![{35F81A43-2C95-4CFD-93CE-EB4255A4E5DF}](https://github.com/user-attachments/assets/96174c03-f297-4fbc-94f9-380c44f79337)
![{27881FD6-6014-45D1-9EEB-FFACD61FD218}](https://github.com/user-attachments/assets/557695a9-8807-4dfd-9af6-7531b69b8ee3)

**Comment by machinewrapped** [comment]
Thanks, I'm not too familiar with RTL languages but it looks like you have a solution that works. I'll see if I can integrate it into the standard flow. Is it just the exported subtitles that have the issue or does RTL text display incorrectly in the GUI too?

**Reply by Theamoz** [reply]
Both the subtitles and the GUI itself have issues, which are effectively resolved with the solution I provided.

**Comment by machinewrapped** [comment]
Which player are you using, by the way?

I think I've got this working - here's a before and after of some translated text in PotPlayer.

Before:
![image](https://github.com/user-attachments/assets/14507f0c-7c02-4d13-bc41-e258ade737e2)

After:
![image](https://github.com/user-attachments/assets/1b72b1de-db92-40ed-9b06-6b2e2dac73be)

It doesn't look completely wrong in either case to my untrained eye, but I guess the quotation mark and period are not where they're supposed to be in the first image?

**Reply by Theamoz** [reply]
I have tried several different players, but none of them have made a difference.
Yeah, you're right, in your example, the only problem is the dots. But if you add an English word to it, the whole sentence gets messed up.

**Comment by machinewrapped** [comment]
Support for adding RTL tags to the output is in v0.8.4:
https://github.com/machinewrapped/gpt-subtrans/releases/tag/v0.8.4

Changes at the GUI level would be a bit more complicated. In the test Arabic translation I did the GUI seems to match how the subtitles are displayed in PotPlayer with the RTL markers added, even with mixed content. If you can provide a sample subtitle which displays incorrectly in the GUI I can look at it in a future release.

**Reply by Theamoz** [reply]
thanks I'm going to test it

**Comment by Theamoz** [comment]
Hey bro, this works really well, awesome 
Also, the GUI is fine, and I was wrong
It’s all good.
